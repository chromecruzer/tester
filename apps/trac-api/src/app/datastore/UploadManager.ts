import util from "util";
import _ from "lodash";
import path from "path";
import {DateTime} from "luxon";
import {UploadConfiguration} from "../ApiConfiguration";
import {
  dateNow,
  supportedUploadFiles,
  UploadMetadata,
  UploadSupportedFileTypes
} from "@trac/datatypes";

const userSessionsFile = 'activeSessions.json';
const uploadMetadataFile = 'uploadFileProperties.json';
import DataStoreUtilities from "./DataStoreUtilities";


const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});


export interface UserSession {
  token: string;
  uploadName: string;
  type: UploadSupportedFileTypes;
  user: string;
  created: Date;
  form: { [id: string]: string };
}

export default class UploadManager {
  public sessions: { [id: string]: UserSession } = {};
  public metadata: UploadMetadata[] = [];
  public uploadPath: string;
  private readonly filePath: (filename) => string;
  private dsu: DataStoreUtilities;

  constructor(private config: UploadConfiguration) {
    this.uploadPath = path.join(...this.config.root);
    this.filePath = filename => path.join(this.uploadPath, filename);
    this.dsu = new DataStoreUtilities();
  }

  public async initialize() {
    const err = [];
    await this.checkUploadDirectories()
      .catch(e => err.push(...e));
    await this.checkFolder('audit_scans');
    await this.loadJson(userSessionsFile)
      .catch(e => console.info('no sessions found'));
    Object.keys(this.sessions).map(async s => {
      const session = this.sessions[s];
      delete this.sessions[s]
      return this.dsu.remove(session.uploadName)
        .catch(e => err.push(new Error(`Cannot remove ${session.uploadName} because ${e.message}`)));
    });
    await this.saveJson(userSessionsFile)
      .catch(e => err.push(e));
    await this.loadJson(uploadMetadataFile)
      .catch(e => console.log('upload archive is empty'));
    if (err.length > 0) {
      return Promise.reject(err);
    }
    console.log('initialize is done');
    // console.log(`archives ${dump(this.metadata)}`)
  }

  private async checkFolder(folder, err = []) {
    const dir = path.join(this.uploadPath, folder);
    console.log(`checking upload folder ${dir}`);
    return this.dsu.mkdir(dir)
      .catch(() =>
        err.push(new Error(`Cannot create upload folder "${folder}" in ${this.uploadPath}`)));
  }

  private async checkUploadDirectories() {
    const err = [];
    supportedUploadFiles().forEach(f => this.checkFolder(f,err));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public async removeSession(token: string, removeFile = false) {
    const err = [];
    const session = this.sessions[token];
    if (session) {
      delete this.sessions[token];
      if (removeFile) {
        await this.dsu.remove(session.uploadName)
          .catch(e => err.push(new Error(`Cannot remove ${session.uploadName} because ${e.message}`)));
      }
      await this.saveJson(userSessionsFile)
        .catch(e => err.push(e));
      if (err.length > 0) {
        return Promise.reject(err);
      }
    }
  }

  public async addSession(filepath, user, type, form) {
    let err = null;
    const result = {
      token: UploadManager.newToken(),
      type,
      uploadName: filepath,
      user,
      created: DateTime.now().toJSDate(),
      form,
    };
    this.sessions[result.token] = result;
    // console.log(`sessions ${dump(this.sessions)}`)
    await this.saveJson(userSessionsFile)
      .catch(e => err = e);
    if (err) {
      return Promise.reject(err);
    }
    return result.token;
  }

  public async archiveSession(token: string) {
    const err = [];
    const session = this.sessions[token] || null;
    if (session === null) {
      return Promise.reject([new Error(`Session with token ${token} does not exist`)]);
    }
    // console.log(`archiving session ${dump(session)}`);
    const archived = {
      name: session.uploadName,
      type: session.type,
      user: session.user,
      date: session.created,
      form: session.form
    };
    this.metadata.push(archived);
    await this.removeSession(token)
      .catch(e => err.push(...e));
    await this.saveJson(uploadMetadataFile)
      .catch(e => err.push(e));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public async markRefresh(user) {
    const err = [];
    const archived = {
      type: 'refresh_data' as UploadSupportedFileTypes,
      user,
      date: dateNow(),
    };
    this.metadata.push(archived);
    await this.saveJson(uploadMetadataFile)
      .catch(e => err.push(e));
    if (err.length > 0) {
      return Promise.reject(err);
    }
  }

  public history() {
    // console.log('setting up history');
    return (req, res) => {
      // console.log('returning history')
      res.json(_.filter(this.metadata, m => m.type !== 'audit_scans'));
    }
  }

  private async loadJson(filename: string) {
    const filepath = this.filePath(filename);
    // console.log(`checking path ${filepath}`);
    if (!await this.dsu.check(filepath)) {
      return;
    }
    // console.log(`reading from path ${filepath}`);
    return this.dsu.retrieve(filepath)
      .then(json => {
        // console.log(`retrieved ${dump(json)}`);
        if (filename === userSessionsFile) {
          this.sessions = json;
        } else {
          this.metadata = json;
        }
      })
      .catch(e => new Error(`File ${filepath} could not be read. ${e.message}`));
  }

  private async saveJson(filename) {
    const filepath = this.filePath(filename);
    const json = filename === userSessionsFile ? this.sessions : this.metadata;
    // console.log(`writing ${dump(json)} to ${filepath}`);
    return this.dsu.store(filepath, json)
  }

  static newToken() {
    return Math.random().toString(36).substring(0, 10)
      .replace('.', '');
  }
}
