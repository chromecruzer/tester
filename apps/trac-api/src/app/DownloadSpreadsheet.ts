import {getDataFields} from "@trac/datatypes";
import _ from "lodash";
import XLSX from 'xlsx';
import UploadManager from "./datastore/UploadManager";
import util from "util";
import {tryCatch} from "@trac/postgresql";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export class DownloadSpreadsheet {

  public convertToSpreadsheet() {
    return tryCatch(async (req, res, next) => {
      const headers = req.body[getDataFields.headers];
      const spreadsheet = req.body[getDataFields.spreadsheet];
      const spreadsheet0 = req.body[getDataFields.spreadsheet0] || null;
      res.attachment(`TracSpreadsheet_${UploadManager.newToken()}.xlsx`);
      res.status(200).end(this.saveFile(this.convertSpreadsheet(headers, spreadsheet), spreadsheet0));
    })
  }

  private convertSpreadsheet(headers, spreadsheet) {
    // console.log(`headers ${headers.length}`);
    return _.map(spreadsheet, r => {
      const result = {};
      for (let h = 0; h < headers.length; h++) {
        // if(headers[h].header === 'Audit Notes') {
        //   console.log(`index ${headers[h].accessor} ${headers[h].header}=${dump(r[headers[h].accessor])}`);
        // }
        result[headers[h].header] = r[headers[h].accessor];
      }
      return result;
    });
  }

  private saveFile(content, sheet0 = null) {
    // console.log(`spreadsheet ${dump(content)}`)
    const workbook = XLSX.utils.book_new();
    let worksheet;
    if (sheet0) {
      worksheet = XLSX.utils.json_to_sheet(sheet0);
      XLSX.utils.book_append_sheet(workbook, worksheet);
    }
    worksheet = XLSX.utils.json_to_sheet(content);
    XLSX.utils.book_append_sheet(workbook, worksheet);
    // XLSX.writeFile(workbook, 'spreadsheet.xlsx', {type: 'buffer', bookType:'xlsx'})
    //  writeFileSync('bufferedspreadsheet.xlsx', XLSX.write(workbook, {type: 'buffer', bookType:'xlsx'}), 'binary');
    return XLSX.write(workbook, {type: 'buffer', bookType: 'xlsx'})
  }
}
