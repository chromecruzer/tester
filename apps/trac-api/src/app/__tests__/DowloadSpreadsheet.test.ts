import UploadManager from "../datastore/UploadManager";
import {DownloadSpreadsheet} from "../DownloadSpreadsheet";

describe('DownloadSpreadsheet', () => {
  let columns, data, saveFileSpy, mockResponse, mockEnd, download;
  beforeEach(() => {
    saveFileSpy = jest.fn().mockImplementation(() => 'converted spreadsheet buffer');
    mockEnd = jest.fn();
    jest.spyOn(UploadManager, 'newToken').mockImplementation(() => 'newToken');
    mockResponse = {
      attachment: jest.fn(),
      status: jest.fn().mockImplementation(() => ({end: mockEnd})),
    }
    columns = [
      {
        Header: 'Text Column',
        accessor: 'textColumn'
      },
      {
        Header: 'Number Column',
        accessor: 'numberColumn'
      },
      {
        Header: 'Date Column',
        accessor: 'dateColumn'
      },
    ];
    data = [
      {
        textColumn: 'text1',
        numberColumn: 102,
        dateColumn: '10/02/21'
      },
      {
        textColumn: 'text2',
        numberColumn: 115,
        dateColumn: '10/02/21'
      },
      {
        textColumn: '3rd text',
        numberColumn: 10.2,
        dateColumn: '10/12/21'
      },
      {
        textColumn: 'text the fourth',
        numberColumn: 12345,
        dateColumn: '10/15/21'
      },
    ];
    download = new DownloadSpreadsheet();
    download.saveFile = saveFileSpy;
  });
  describe('convertToSpreadsheet', () => {
    it('should create a spreadsheet from headers and data', () => {
      const convert = download.convertToSpreadsheet();
      return convert({
        body: {
          headers: columns,
          spreadsheet: data,
        }
      }, mockResponse, jest.fn())
        .then(() => {
          expect(saveFileSpy.mock.calls).toMatchSnapshot();
          expect(mockResponse.attachment.mock.calls).toMatchSnapshot();
        });
    });
  });
});
