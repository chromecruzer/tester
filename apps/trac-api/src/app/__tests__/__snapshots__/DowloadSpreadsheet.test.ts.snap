// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadSpreadsheet convertToSpreadsheet should create a spreadsheet from headers and data 1`] = `
Array [
  Array [
    Array [
      Object {
        "Date Column": "10/02/21",
        "Number Column": 102,
        "Text Column": "text1",
      },
      Object {
        "Date Column": "10/02/21",
        "Number Column": 115,
        "Text Column": "text2",
      },
      Object {
        "Date Column": "10/12/21",
        "Number Column": 10.2,
        "Text Column": "3rd text",
      },
      Object {
        "Date Column": "10/15/21",
        "Number Column": 12345,
        "Text Column": "text the fourth",
      },
    ],
  ],
]
`;

exports[`DownloadSpreadsheet convertToSpreadsheet should create a spreadsheet from headers and data 2`] = `
Array [
  Array [
    "TracSpreadsheet_newToken.xlsx",
  ],
]
`;
